File: p1/gth773s/java/src/abcd/Memo.java
package abcd;

import com.google.common.base.Function;
import com.google.common.base.Joiner;
import com.google.common.collect.Lists;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;

import static com.google.common.collect.Lists.newArrayList;
import static com.google.common.collect.Maps.newHashMap;

class Memo {

    private Map<Diff, LBound> map = newHashMap();

    L get(Inequality inequality) {
        Diff diff = inequality.diff();
        LBound bound = diff.isConstant() ? diff.fixedBound() : map.get(diff);
        return bound == null ? null : bound.get(inequality.c());
    }

    void modify(LBound.Action action, Inequality inequality, L l) {
        Diff diff = inequality.diff();
        LBound bound = map.get(diff);
        if (bound == null) map.put(diff, bound = new LBound());
        bound.modify(action, l, inequality.c());
        if (bound.isVacuous()) map.remove(diff);
    }

    @Override
    public String toString() {
        List<Diff> keys = newArrayList(map.keySet());
        Collections.sort(keys, valuePairComparator);
        return Joiner.on("\n").join(Lists.transform(keys, valuePairToString));
    }

    private Function<Diff, String> valuePairToString = new Function<Diff, String>() {
        public String apply(Diff diff) {
            return map.get(diff).toString(diff);
        }
    };

    private static Comparator<Diff> valuePairComparator = new Comparator<Diff>() {
        public int compare(Diff a, Diff b) {
            int result = compare(a.b(), b.b());
            return result != 0 ? result : compare(a.a(), b.a());
        }
        private int compare(Value a, Value b) {
            if (a == b) return 0;
            if (a.isConstant() != b.isConstant())
                return Boolean.valueOf(a.isConstant())
                    .compareTo(Boolean.valueOf(b.isConstant()));
            else if (a.isConstant())
                return Integer.valueOf(a.constant).compareTo(Integer.valueOf(b.constant));
            else
                return a.nick.compareTo(b.nick);
        }
    };

}

File: p1/gth773s/phase1-gth773s/Memo.cpp
#include "PhaseOne.h"

using namespace llvm;
using namespace std;

namespace phaseone {

  L Memo::get(Inequality inequality) {
    Diff diff = inequality.diff();
    APInt c = inequality.c();
    L l = diff.fixedBoundGet(c);
    if (l != L_NULL) return l;
    if (memo.find(diff) == memo.end()) return L_NULL;
    return memo[diff]->get(c);
  }

  void Memo::modify(bool strengthen, Inequality inequality, L l) {
    Diff diff = inequality.diff();
    APInt c = inequality.c();
    LBound *bound;
    if (memo.find(diff) == memo.end()) {
      bound = LBound::create();
      memo[diff] = bound;
    } else {
      bound = memo[diff];
    }
    bound->modify(strengthen, l, c);
    if (bound->isVacuous()) memo.erase(diff);
  }

}
